### PR TITLE
Additional options for EventGhost.TriggerEvent

### PR DESCRIPTION
**Pull request by kdschlosser**
queueEvent:  
If the wait time is set to 0.00 this option will appear. Adding the
event to the queue means that the event will get triggered after the
event that caused this action has processed all of it's macros and also
after processing any events that have come in while the event that
triggered this action was running.

  If unchecked the event will get triggered right away Not being added
to the queue and not waiting until the event that started this action
has finished processing.

restoreEvent:
  If add to queue has not been checked this option will appear. This
relates more to the scripting portions of EventGhost. What this done is
each and every time an event gets triggered there are 2 variables that
get set into place. Those 2 variables are eg.event and eg.eventString.
When you trigger an event while the current event is running those 2
variables will get changed to the new event. Upon completion of the new
event if you would like to change those variables back to the event that
ran this action then check this box.

### CODE REVIEW

**Comment by kdschlosser** [comment]
This is for the Trigger Event Action and not eg.TriggerEvent

File: plugins/EventGhost/__init__.py [hunk 1]
```diff
@@ -608,20 +644,75 @@ class text:
...
+
+        queueEventCtrl.SetValue(queueEvent)
+        restoreEventCtrl.SetValue(restoreEvent)
+        queueEventText = panel.StaticText(text.text4)
+        restoreEventText = panel.StaticText(text.text5)
+
+        def on_spin(evt):
+            def check_spin():
+                value = bool(waitTimeCtrl.GetValue())
+                queueEventText.Show(not value)
```
Comment (topic2k) [inline plugins/EventGhost/__init__.py]: I would preferr to disable the options instead of hiding them.

File: plugins/EventGhost/__init__.py [hunk 2]
```diff
@@ -608,20 +644,75 @@ class text:
...
+                queueEventCtrl.Show(not value)
+                restoreEventText.Show(
+                    not value or (not value and not queueEventCtrl.GetValue())
+                )
+                restoreEventCtrl.Show(
+                    not value or (not value and not queueEventCtrl.GetValue())
+                )
+            wx.CallLater(20, check_spin).Start()
```
Comment (topic2k) [inline plugins/EventGhost/__init__.py]: What does the ``.Start()` do?

File: plugins/EventGhost/__init__.py [hunk 3]
```diff
@@ -632,18 +723,71 @@ def Configure(self, eventString="", waitTime=0):
             (waitTimeCtrl, 0, wx.ALL, 5),
             (panel.StaticText(text.text3), 0, wx.ALIGN_CENTER_VERTICAL),
         )
-        panel.sizer.AddMany(((sizer1, 0, wx.EXPAND), (sizer2, 0, wx.EXPAND)))
+
+        sizer3 = eg.HBoxSizer(
```
Comment (topic2k) [inline plugins/EventGhost/__init__.py]: optional: replace sizer3 and sizer4 with just ```        panel.AddLine(queueEventText, queueEventCtrl)
        panel.AddLine(restoreEventText, restoreEventCtrl)
```
It's a little shorter and the checkboxes are aligned.

File: plugins/EventGhost/__init__.py (hunk 2 above)
Comment (kdschlosser) [reply plugins/EventGhost/__init__.py]: I have gotten into the habit of using call later whenever i am checking a wx widget for a value that can be changed by keyboard input as an example if you have a TextCtrl that you want to make sure it is only digits that are entered and you want to not allow the input if it's not. I know this can be done on a per character basis which i feel is more complex trying to use keycodes and what have you to determine this. so if you bind to EVT_CHAR or EVT_CHARHOOK if you grab the value of the control to be able to parse it it will not reflect the current keyboard input that triggered the event. but if you use a CallLater which is basically the same thing as CallAfter except it has a timer for when it should run you need to use Start() to start the timer. but you can call a function that would check and parse the input as needed and it would now be updated so when you call GetValue() on the control it will have been updated with the latest keystrokes.

File: plugins/EventGhost/__init__.py (hunk 3 above)
Comment (kdschlosser) [reply plugins/EventGhost/__init__.py]: I do agree with you on this. and this is one of those things i really really wanted to fix. along with the eg.EqualizeWidths. But i do have a nice formatter that I use in all of my plugins that I have found does a fantastic job of lining things up. I am not a fan of the AddLine because It leaves a rather large gap between the lines. I just wanted to keep it the same as it had already been coded.

File: plugins/EventGhost/__init__.py (hunk 1 above)
Comment (kdschlosser) [reply plugins/EventGhost/__init__.py]: ok No problem there either.

File: plugins/EventGhost/__init__.py (hunk 2 above)
Comment (topic2k) [reply plugins/EventGhost/__init__.py]: You don't need to add Start() to CallLater(). In CallLater's __init__ method Start() will automatically be called.

Comment (kdschlosser) [reply plugins/EventGhost/__init__.py]: really?!?! this is did not know.

File: plugins/EventGhost/__init__.py [hunk 4]
```diff
@@ -595,55 +595,292 @@ def __call__(self):
...
+    ):
+
+        def parse(value):
+            if value is None:
+                return None
+            parsed_value = eg.ParseString(value)
+            if value == parsed_value:
+                try:
+                    value = eval(value)
+                except SyntaxError:
```
Comment (topic2k) [inline plugins/EventGhost/__init__.py]: Here i get a NameError. With `except (SyntaxError, NameError):` it looks good to me.

File: plugins/EventGhost/__init__.py [hunk 5]
```diff
@@ -595,55 +595,292 @@ def __call__(self):
...
-            (panel.StaticText(text.text2), 0, wx.ALIGN_CENTER_VERTICAL),
-            (waitTimeCtrl, 0, wx.ALL, 5),
-            (panel.StaticText(text.text3), 0, wx.ALIGN_CENTER_VERTICAL),
-        )
-        panel.sizer.AddMany(((sizer1, 0, wx.EXPAND), (sizer2, 0, wx.EXPAND)))
         while panel.Affirmed():
             panel.SetResult(
                 eventStringCtrl.GetValue(),
                 waitTimeCtrl.GetValue(),
+                payloadCtrl.GetValue() if payloadCtrl.GetValue() else None,
```
Comment (topic2k) [inline plugins/EventGhost/__init__.py]: replace `None` with `''` else you get an exception next time you wan't to edit the action.

Comment (kdschlosser) [reply plugins/EventGhost/__init__.py]: Payload has to be sent in as None if there isn't one. this is because it is the default value for TriggerEvent. So it makes it easier in the __call__ method if it can simply pass the None. so what I did was I just added a check for the None and to change it to ""

**Comment by kdschlosser** [comment]
All fixed lil' buddy

**Comment by kdschlosser** [comment]
This has an error in it..

You didn't test it :(

**Comment by kdschlosser** [comment]
sorry my mistake on this one As it turns out it is working. something changed in my development copy of EG and i thought that i sent the typo in the PR but i didn't

**Comment by per1234** [comment]
A really significant breaking change was introduced by https://github.com/EventGhost/EventGhost/pull/279/commits/294f16bf838cb58962cec90d7628545293343c6a. I did remember this being mentioned in a bit more detail either in conversations here or on the forum but now I don't find it.

The documentation update is well written but the change is undocumented in the commit history and I fear this will propagate to catching the users unaware at the time of the next release.

The commit message could have been more clear about this but that wouldn't have mattered because it was lost anyway when it was squashed in with unrelated commits during the merge. It's not at all obvious to me that ["Additional options for EventGhost.TriggerEvent"](https://github.com/EventGhost/EventGhost/commit/4c1ac340fc3e09131960fefed066484bddfd2e75) would include a significant change in the previous behavior of the plugin.

I do understand the reasoning behind this change but it has thoroughly broken my tree and I'm sure it will do so for many other users. A change like this should:
- Have a dedicated commit which clearly explains the breakage and defends the need for it. In the interest of that, it's better to submit a dedicated PR for each distinct change rather than lumping only slightly related commits into a PR. Often changes made after review of the PR require squashing it during the merge to keep the commit history clean. This will also keep the discussion on each PR focused. This also makes it easy to roll back through changes while troubleshooting.
- Be clearly documented in the release notes. I'm concerned that the lack of the above could result in no mention of this in the release notes.
- Result in a major version bump so that users will know that they will need to dedicate some time to thoroughly investigate how it affects them and deal with any problems caused by updating to the next EventGhost release.

**Comment by kdschlosser** [comment]
@per1234
This PR shouldn't have broken anything at all.

There are defaulted values set in place for the action.

I strive to not have anyone's tree get broken from changes we make. So please explain in detail what exactly happened. I want to fix/update this in a manner that will not break API or anyone's tree for that matter.

Thanks again
K

**Comment by kdschlosser** [comment]
@per1234 
Never mind. I see where it got broken.

With how I handled the event string. if there is a . in it the Main prefix goes away. My dumbass didn't think. LOL. I can add a checkbox to enable or disable the Main prefix and have it set to off as a default.

I personally hated the thing using the Main. all the time. This was on EG 0.4 because the '*' for wild carding only worked in specific places. I think this caused me issues. But the user should be able to supply their own Prefix. maybe the better way to handle it is to add a prefix text ctrl and a suffix text ctrl. I think that would be a far better solution.


so basically the __call__ would be like so.

    def __call__(
        suffix,
        waitTime=0,
        prefix='Main',
        payload=None,
        queueEvent=True,
        restoreEvent=False
    ):


This I believe would not break the API at all. and I can change the parameter name because the EG xml file uses positional args and not keywords. so long as i don't change their place we should be good to go.

I am assuming this is the problem you had.

**Comment by per1234** [comment]
Yes. I have a ton of things like:
```
eg.plugins.EventGhost.TriggerEvent(u'relay.1on', 0.0)
```
Then the macro trigger is:
```
Main.relay.1on
```
Which has worked as long as I've been using EventGhost.

I think that it could definitely be considered a bug that EventGhost used to add `Main.` in this case but it's such a long standing bug that many people's tree's will be coded to expect it. I'm not necessarily arguing against the change but I think your idea for preventing breakage by adding the option and making it default to the old behavior is a good idea. It's not a huge problem for me since I had some recollection of the change being mentioned and so I figured out the cause of the breakage pretty quickly  but for others it could definitely cause a lot of confusion. I was actually thrown off for a little while because I couldn't figure out why some of my TriggerEvents were still working as before (the ones without .) because I didn't remember the details of the change. My concern was more that I had a hard time tracking down the commit that caused it, even though I knew about where to look, and suspected this would slip through the cracks when the release notes were written.

**Comment by kdschlosser** [comment]
I didn't even think about the implication of Main being gone if there is a . in the event name. I am going to add a prefix and suffix text control. this way the user will be able to go through and change the event being generated if they want. but by default it will add the Main so not to break anyone's trees. 

Since you have been using a version of EG with this addition have you tried any of the other features it offers? like running the event immediately?

**Comment by per1234** [comment]
I rolled back to the last build before this was merged because I needed a working tree and didn't have the time to update all the triggers.

I didn't have the latest build installed for long enough to check out any new features. I'll give them a try once I get back to the latest version again and let you know how it goes.

I just got a new (to me) computer and went from Windows 7 to 10 on both my computers at the same time so I'm dealing with a lot of system changes all at once right now but starting to get everything back to working condition.

**Comment by kdschlosser** [comment]
why would you do such a horrible thing....

MY GOD!

Hell if I had a choice I would probably be running Windows 2000 most stable OS Microsoft has ever made. Windows 7 is a very very close second.

To be honest with as much modifying of Windows 7 as I have done it probably beats out 2000. I am running what could be considered Windows 7 Thin PC x64 Microsoft didn't make such a thing but I have stripped out all of the crap from it. has under a 10 gig footprint and uses under a gig of RAM.

the installation of Windows 10 has caused you to not own a thing on your computer. they do whatever it is they want when they want. I do not know if anyone has found a way to shut off all of the BS that Windows 10 does without your permission. Or even if it asks it will simply ignore you. Personally I do not like the fact that they lock out the ability to do various things  as an example: Install unsigned drivers - a warning should be left and the user makes the decision. What makes it worse is they remove it then tell everyone that it will only get removed if it is a clean install and not an upgrade/update. LIES!

Whenever you want to reclaim your machine speed and resources Let me know.

If you would be willing... Can you tell me if they fixed USB redirection when using RDP? Because it was working in an RC but when the full release took place they broke it somehow.. Last i checked (about 6 months ago) it was still broken.
